### PR TITLE
HPCC-18515 Dafilesrv remote read streaming

### PR DESCRIPTION
**Pull request by jakesmith**
Add support for dynamic remote disk read projects to dafilesrv
(no filtering yet), via JSON requests.

Output format in either binary or JSON.
Continuation either from same instance or if unknown handle,
challenge/response, so client can resend original with cursor.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

This PR includes changes to testsocket to test the functionality, that includes an option [-simulate-rrresend], that tests continuation of an unassociated cursor (e.g. if dafilesrv has been restarted/connection was lost).

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - for comment.

**Comment by AttilaVamos** [comment]
@jakesmith it seems (based on the failed test cases) Gavin's fix is missing from your master branch.

**Comment by jakesmith** [comment]
Just double-checked, it is up to date with latest master branch.
Master branch SHA = 4bf4bdf9a85d6afa59fd99004e13799a6a07911f
This PR HEAD^ = 4bf4bdf9a85d6afa59fd99004e13799a6a07911f

@AttilaVamos - Which fixes of Gavin's did you mean?

File: rtl/eclrtl/rtlrecord.cpp [hunk 1]
```diff
@@ -591,7 +591,9 @@ ISourceRowPrefetcher * COutputMetaData::createDiskPrefetcher(ICodeContext * ctx,
 //  Should do something like this, if the flag existed...
 //  if (!getMetaFlags() & MDFhasaliendatatype_or_ifblock)
 //      return new CDefaultPrefetcher(queryRecordAccessor(true));
-    return new CSimpleSourceRowPrefetcher(*this, ctx, activityId);
+
+    //return new CSimpleSourceRowPrefetcher(*this, ctx, activityId);
+    return new CDefaultPrefetcher(queryRecordAccessor(true));
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlrecord.cpp]: This is why you are getting the smoketest failures

**Comment by AttilaVamos** [comment]
@jakesmith Sorry I was wrong.

**Comment by richardkchapman** [comment]
Nah, Jake was - he deliberately rolled back the change in question! This can't be merged until that issues is properly addressed (but it's only for comment at present)

**Comment by jakesmith** [comment]
>Nah, Jake was - he deliberately rolled back the change in question!

ah that change, yes - deliberate - not for merging yet.

**Comment by jakesmith** [comment]
Rebased now that Gavin's change is in, and now passes smoke tests.

@richardkchapman , @ghalliday - please review.

**Comment by ghalliday** [comment]
@jakesmith check boxes aren't checked updated

**Comment by jakesmith** [comment]
@ghalliday - check boxes done.

NB: compressed file support is not implemented in this PR.

File: common/remote/sockfile.cpp [hunk 2]
```diff
@@ -3811,17 +3831,244 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
...
+        {
+            while (!eofseen && ((chooseN == 0) || (processed < chooseN)))
+            {
+                while (!prefetchBuffer.eos())
+                {
+                    prefetcher->readAhead(prefetchBuffer);
+                    const byte * next = prefetchBuffer.queryRow();
+                    size32_t sizeRead = prefetchBuffer.queryRowSize();
+                    if (segMonitorsMatch(next))
+                        rowSz = helper->transform(*outBuilder, next);
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: very minor: more efficient to assign to a local, and then assign to the reference at the point it returns.

**Review by ghalliday — COMMENTED** [review]
The PR looks broadly good.  There a a few minor comments and questions.

There are a few other bigger issues, but they could go into separate jiras instead, so that this can be merged as a first version.  Main issues are:
* Helper creation is in the wrong place
* Support for grouped datasets
* Code should be in a separate file from sockfile.cpp - it is likely to be used in some other places.
* The code that was moved in commit #2 could do with reviewing - there are various issues I would have commented on if it was new code.

Then there are also other issues which probably already have jiras - compressed input and output and other issues.

**Comment by ghalliday** [comment]
@jakesmith see summary comment.

File: common/remote/sockfile.cpp [hunk 3]
```diff
@@ -3811,17 +3831,244 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
     }
 };
 
+interface IRemoteReadCursor : extends IInterface
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/remote/sockfile.cpp]: Agreed, will merge into IRemoteActivity.

File: common/remote/sockfile.cpp [hunk 4]
```diff
@@ -3811,17 +3831,244 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
...
+        const RtlRecordTypeInfo *inputRecordTypeInfo = getTypeInfo(*inputFieldInfoDeserializer, actNode, "input"); // NB: inputFieldInfoDeserializer owns returned object
+        const RtlRecordTypeInfo *outputRecordTypeInfo = getTypeInfo(*outputFieldInfoDeserializer, actNode, "output"); // NB: outputFieldInfoDeserializer owns returned object
+
+        outMeta.setown(new CDynamicOutputMetaData(*outputRecordTypeInfo));
+
+        unsigned __int64 chooseN = actNode.getPropInt64("choosen", defaultFileStreamChooseN);
+        unsigned __int64 skipN = actNode.getPropInt64("skipN", defaultFileStreamSkipN);
+        unsigned __int64 rowLimit = actNode.getPropInt64("rowLimit", defaultFileStreamRowLimit);
+        Owned<IOutputMetaData> in = new CDynamicOutputMetaData(*inputRecordTypeInfo);
+        helper.setown(createDiskReadArg(fileName, in.getClear(), outMeta.getLink(), chooseN, skipN, rowLimit));
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/remote/sockfile.cpp]: right, I'll move out to factory method.

File: common/remote/sockfile.cpp [hunk 5]
```diff
@@ -3811,17 +3831,244 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
...
+    {
+        if (outBuilder)
+            delete outBuilder;
+    }
+    virtual const void *nextRow(size32_t &rowSz) override
+    {
+        if (!opened) open();
+        if (needTransform)
+        {
+            while (!eofseen && ((chooseN == 0) || (processed < chooseN)))
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/remote/sockfile.cpp]: Yes, 0 means all records (code same as in engines)

File: common/remote/sockfile.cpp [hunk 6]
```diff
@@ -3811,17 +3831,244 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
...
+                }
+                eofseen = true;
+            }
+        }
+        close();
+        rowSz = 0;
+        return nullptr;
+    }
+// IRemoteActivity impl.
+    virtual void serialize(MemoryBuffer &tgt) const override
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/remote/sockfile.cpp]: will change (and restore() to restoreCursor())

File: common/remote/sockfile.cpp [hunk 7]
```diff
@@ -3811,17 +3831,244 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
...
+        tgt.append(processed);
+        delayed.write();
+    }
+    virtual void restore(MemoryBuffer &src) override
+    {
+        unsigned len;
+        src.read(len);
+        src.read(startPos);
+        src.read(processed);
+    }
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/remote/sockfile.cpp]: not sure it (len) serves any useful purpose, I'll remove len.

File: common/remote/sockfile.cpp [hunk 8]
```diff
@@ -5637,6 +5884,185 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
...
+        else if (strieq("json", outputFmtStr))
+            outputFormat = outFmt_Json;
+        else
+            outputFormat = outFmt_Binary;
+        if (outFmt_Binary != outputFormat)
+            responseTree.setown(createPTree("Response"));
+
+        Owned<IRemoteActivity> outputActivity;
+        OpenFileInfo fileInfo;
+        if (!cursorHandle)
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/remote/sockfile.cpp]: 0 was intentional. Dafilesrv handles are 1 based.

File: tools/testsocket/testsocket.cpp [hunk 9]
```diff
@@ -515,17 +615,22 @@ int doSendQuery(const char * ip, unsigned port, const char * base)
     }
 
     const char * query = fullQuery.str();
-    int len=strlen(query);
-    int sendlen = len;
+    size32_t queryLen=strlen(query);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply tools/testsocket/testsocket.cpp]: will add

File: common/remote/sockfile.cpp [hunk 10]
```diff
@@ -3811,17 +3831,244 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
...
+        if (!opened) open();
+        if (needTransform)
+        {
+            while (!eofseen && ((chooseN == 0) || (processed < chooseN)))
+            {
+                while (!prefetchBuffer.eos())
+                {
+                    prefetcher->readAhead(prefetchBuffer);
+                    const byte * next = prefetchBuffer.queryRow();
+                    size32_t sizeRead = prefetchBuffer.queryRowSize();
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/remote/sockfile.cpp]: will remove

File: common/remote/sockfile.cpp [hunk 11]
```diff
@@ -3811,17 +3832,247 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
...
+        // TBD
+    }
+    bool segMonitorsMatch(const void *row) { return true; }
+public:
+    CRemoteDiskReadActivity(IHThorDiskReadArg &_helper, IRtlFieldTypeDeserializer &_inputFieldInfoDeserializer, IRtlFieldTypeDeserializer &_outputFieldInfoDeserializer)
+        : helper(&_helper), inputFieldInfoDeserializer(&_inputFieldInfoDeserializer), outputFieldInfoDeserializer(&_outputFieldInfoDeserializer), prefetchBuffer(nullptr)
+    {
+        /* NB: this objects links the IRtlFieldTypeDeserializer's, purely because RtlRecordTypeInfo that have been deserialized are used
+         * by this class and the helper, and would become invalid if the deserializer objects were destroyed.
+         */
```
Comment (jakesmith) [inline common/remote/sockfile.cpp]: @ghalliday - is there a cleaner alternative ? Seems a shame to have these members, solely for this reason.

As I mentioned before, to me it looks like getTypeInfo() that get a RtlRecordTypeInfo info from the IRtlFieldTypeDeserializer would be safe to use beyond releasing the IRtlFieldTypeDeserializer, e.g. if RtlRecordTypeInfo objects linked the IRtlFieldTypeDeserializer, then this wouldn't be necessary.

File: common/remote/sockfile.cpp [hunk 12]
```diff
@@ -3811,17 +3831,244 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
...
+    if (inputJson)
+        fieldInfo = fieldTypeDeserializer.deserialize(*inputJson);
+    else
+    {
+        StringBuffer binTypePropName(typePropName);
+        MemoryBuffer mb;
+        actNode.getPropBin(binTypePropName.append("Bin").str(), mb);
+        fieldInfo = fieldTypeDeserializer.deserialize(mb);
+    }
+    return static_cast<const RtlRecordTypeInfo *>(fieldInfo);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/remote/sockfile.cpp]: do you mean assertex(!fieldInfo.isScalar()) ?

File: common/remote/sockfile.cpp [hunk 13]
```diff
@@ -5637,6 +5884,185 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
...
+        {
+            IOutputMetaData *out = outputActivity->queryOutputMeta();
+            unsigned __int64 initProcessed = outputActivity->queryProcessed();
+            if (outFmt_Binary == outputFormat)
+            {
+                DelayedSizeMarker dataLenMarker(reply); // data length
+                for (unsigned __int64 i=0; i<defaultDaFSNumRecs; i++)
+                {
+                    size32_t rowSz;
+                    const void *row = outputActivity->nextRow(rowSz);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/remote/sockfile.cpp]: Opened separate sub-task JIRA: HPCC-18568

File: common/remote/sockfile.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/remote/sockfile.cpp]: I've added:
```
    assertex(fieldInfo->getType() == type_record);
```

File: common/remote/sockfile.cpp (hunk 2 above)
Comment (jakesmith) [reply common/remote/sockfile.cpp]: will add local var.

File: common/remote/sockfile.cpp (hunk 11 above)
Comment (ghalliday) [reply common/remote/sockfile.cpp]: The disk read activity shouldn't know anything about the field information deserializer.  That should be manged by the helper.  Possibly needs a derived version, or a query object.

File: common/remote/sockfile.cpp [hunk 14]
```diff
@@ -5637,6 +5884,185 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
...
+                        {
+                            UNIMPLEMENTED;
+                            break;
+                        }
+                        default:
+                            throwUnexpected();
+                    }
+                }
+            }
+            if (initProcessed < outputActivity->queryProcessed())
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/remote/sockfile.cpp]: Yes that would be better, have changed.

File: common/remote/sockfile.cpp (hunk 11 above)
Comment (jakesmith) [reply common/remote/sockfile.cpp]: I've moved the deserializers into a variety of the helper.

**Comment by jakesmith** [comment]
@ghalliday - please review changes.

File: common/remote/sockfile.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ghalliday) [reply common/remote/sockfile.cpp]: choosen(0) used to mean all records, but it was changed to mean 0.  I wonder if that means we have some lurking problems in our code...

File: common/remote/sockfile.cpp [hunk 15]
```diff
@@ -3811,17 +3832,228 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
...
+    }
+    virtual unsigned __int64 queryProcessed() const override
+    {
+        return processed;
+    }
+    virtual IOutputMetaData *queryOutputMeta() const override
+    {
+        return outMeta;
+    }
+    virtual StringBuffer &getInfoStr(StringBuffer &out) const
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: missing override

File: common/remote/sockfile.cpp [hunk 16]
```diff
@@ -5637,6 +5869,179 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
...
+            outputActivity->getInfoStr(requestStr);
+            Owned<StringAttrItem> name = new StringAttrItem(requestStr);
+
+            CriticalBlock block(sect);
+            cursorHandle = getNextHandle();
+            client.previdx = client.openFiles.ordinality();
+            client.openFiles.append(OpenFileInfo(cursorHandle, outputActivity, name));
+        }
+        else if (!lookupFileIOHandle(cursorHandle, fileInfo))
+            cursorHandle = NotFound; // challenge response ..
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: should this return 0 for consistency with above.

File: common/remote/sockfile.cpp [hunk 17]
```diff
@@ -5637,6 +5869,179 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
...
+        else // known handle, continuation
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: What happens if the same continuation request is sent to dafilesrv?  E.g., if the cursor is returned to esp, which then uses it on a subsequent call, and when refreshing the window?

File: common/remote/sockfile.cpp [hunk 18]
```diff
@@ -5637,6 +5869,179 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
...
+            outputActivity.set(fileInfo.activity);
+
+        if (outFmt_Binary != outputFormat)
+            responseTree->setPropInt("cursor", cursorHandle);
+        else
+            reply.append(cursorHandle);
+        if (NotFound != cursorHandle)
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: again, here cursorHandle is checked against NotFound instead of 0.

**Review by ghalliday — COMMENTED** [review]
A few more comments

**Comment by ghalliday** [comment]
@jakesmith a couple more comments, but very close to being able to merge

Do we have jiras for 
- refactoring the code to move it out of sockfile.cpp
- support for skipping.  (Will need to return the number left to skip + possibly optimize if can avoiding opening the file?
- finishing reading on compressed file block boundaries

File: common/remote/sockfile.cpp (hunk 17 above)
Comment (jakesmith) [reply common/remote/sockfile.cpp]: Currently the test client (testsocket) does not send the serialized cursor continuation, 
*unless* it gets a challenge response from dafilesrv, at which point it makes another request with the full request + the cursor, so a new dafilesrv can continue.

For the example you give, esp could always _not_ send in a cursor handle and always send in the cursor. That would have the effect of allowing refresh and continuation.

However, I think it's better that if any client does send a cursor, dafilesrv uses it.
That way a client with a handle can also send the cursor if it wants to (e.g. on a refresh), or a client can send the handle only (and be challenged if unknown) and continue to stream.

File: common/remote/sockfile.cpp (hunk 16 above)
Comment (jakesmith) [reply common/remote/sockfile.cpp]: could be, I'll change

File: common/remote/sockfile.cpp (hunk 15 above)
Comment (jakesmith) [reply common/remote/sockfile.cpp]: Will add

File: common/remote/sockfile.cpp (hunk 18 above)
Comment (jakesmith) [reply common/remote/sockfile.cpp]: will change use of NotFound to 0 for consistency.

**Comment by jakesmith** [comment]
@ghalliday - please see last commit.

I have opened sub-issues up on the 3 issues above.

File: common/remote/sockfile.cpp [hunk 19]
```diff
@@ -3811,17 +3832,243 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
...
+    Linked<IOutputMetaData> outMeta;
+    bool cursorDirty = false;
+
+    void checkOpen()
+    {
+        if (opened)
+        {
+            if (!cursorDirty)
+                return;
+            if (prefetchBuffer.tell() != startPos)
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: this also needs to reset eofseen.  That probably requires the result from canMatchAny() to be stored separately

File: common/remote/sockfile.cpp [hunk 20]
```diff
@@ -3811,17 +3832,243 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
...
+    MemoryBuffer resultBuffer;
+    MemoryBufferBuilder *outBuilder = nullptr;
+    unsigned __int64 limit = 0;
+    unsigned __int64 chooseN = 0;
+    IArrayOf<IKeySegmentMonitor> segMonitors;
+    Owned<ISourceRowPrefetcher> prefetcher;
+    CThorContiguousRowBuffer prefetchBuffer;
+    bool opened = false;
+    bool eofseen = false;
+    bool needTransform = true;
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: minor: fields could be grouped more logically, or all booleans together to improve packing.

**Review by ghalliday — COMMENTED** [review]
Two minor comments.

**Comment by ghalliday** [comment]
@jakesmith

**Comment by jakesmith** [comment]
@ghalliday - squashed

**Pull request merged**